Hadar Mentel 205642366 
Linoy Halifa 201052198
Noam Arian 311271829

In [9]:
# Question A: Fit a logistic regression model that uses income and balance to predict default.


import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split



# Load and preprocess the dataset
default_data = pd.read_csv("Default.csv")
default_data['default'] = default_data['default'].map({'Yes': 1, 'No': 0})
default_data['student'] = default_data['student'].map({'Yes': 1, 'No': 0})


# Fit logistic regression model using the entire dataset
X = default_data[['income', 'balance']]  # Predictors
y = default_data['default']  # Response

logit_model = LogisticRegression()
logit_model.fit(X, y)

# Print model coefficients
print("Intercept (β0):", logit_model.intercept_[0])
print("Coefficient for income (β1):", logit_model.coef_[0][0])
print("Coefficient for balance (β2):", logit_model.coef_[0][1])

Intercept (β0): -11.54046791864473
Coefficient for income (β1): 2.0808974142865492e-05
Coefficient for balance (β2): 0.00564710264855791


Question B:
Using the validation set approach, estimate the test error of this
model. In order to do this, you must perform the following steps:
i. Split the sample set into a training set and a validation set.

ii. Fit a multiple logistic regression model using only the training
observations.

iii. Obtain a prediction of default status for each individual in
the validation set by computing the posterior probability of
default for that individual, and classifying the individual to
the default category if the posterior probability is greater
than 0.5.

iv. Compute the validation set error, which is the fraction of
the observations in the validation set that are misclassified.

In [10]:
# Question B: 


# Split data into predictors (income, balance) and response (default)
X = default_data[['income', 'balance']]
y = default_data['default']

# Split into training (50%) and validation (50%) sets
random_seed = 44
#random_seed = 84
#random_seed = 124
X_train, X_validation, y_train, y_validation = train_test_split(X, y, test_size=0.5, random_state=random_seed)

# Fit logistic regression model
logit_model = LogisticRegression()
logit_model.fit(X_train, y_train)

# Print coefficients
print("Intercept (β0):", logit_model.intercept_[0])
print("Coefficient for income (β1):", logit_model.coef_[0][0])
print("Coefficient for balance (β2):", logit_model.coef_[0][1])

Intercept (β0): -11.539641735299313
Coefficient for income (β1): 2.3126098159838207e-05
Coefficient for balance (β2): 0.00559445780686294


In [6]:
y_prob = logit_model.predict_proba(X_validation)[:, 1]  # Probability for "default = 1"

# Classify as "default" (1) if posterior probability > 0.5
y_pred = (y_prob > 0.5).astype(int)

# Display the results (optional: if you want to see probabilities and classifications)
for i in range(len(y_validation)):
    print(f"True Label: {y_validation.iloc[i]}, Probability of Default: {y_prob[i]:.4f}, Predicted Class: {y_pred[i]}")

True Label: 0, Probability of Default: 0.0000, Predicted Class: 0
True Label: 0, Probability of Default: 0.0320, Predicted Class: 0
True Label: 0, Probability of Default: 0.0001, Predicted Class: 0
True Label: 0, Probability of Default: 0.0003, Predicted Class: 0
True Label: 0, Probability of Default: 0.0011, Predicted Class: 0
True Label: 0, Probability of Default: 0.0012, Predicted Class: 0
True Label: 0, Probability of Default: 0.0065, Predicted Class: 0
True Label: 0, Probability of Default: 0.0043, Predicted Class: 0
True Label: 0, Probability of Default: 0.0004, Predicted Class: 0
True Label: 0, Probability of Default: 0.0461, Predicted Class: 0
True Label: 0, Probability of Default: 0.0055, Predicted Class: 0
True Label: 0, Probability of Default: 0.2210, Predicted Class: 0
True Label: 0, Probability of Default: 0.0001, Predicted Class: 0
True Label: 0, Probability of Default: 0.0041, Predicted Class: 0
True Label: 0, Probability of Default: 0.0001, Predicted Class: 0
True Label

In [7]:
# Calculate the number of misclassified observations
num_misclassified = (y_validation != y_pred).sum()

# Calculate the total number of observations in the validation set
total_observations = len(y_validation)

# Compute the validation set error
validation_error = num_misclassified / total_observations

print("Validation Set Error:", validation_error)

Validation Set Error: 0.0252


 Question C: Repeat the process in (b) three times, using three different splits of the observations into a training set and a validation set. Comment on the results obtained.

# Question C: 
Switching seeds creates different random splits of the data into training and validation sets. Each seed shuffles the data in a unique way, ensuring the splits are not the same. This helps test how consistent the model's performance is across different subsets of the data.

random_seed = 44:

- Intercept (β0): -11.539641735343633
- 
Coefficient for income (β1): 2.3126098159743587e-0
- 
Coefficient for balance (β2): 0.0055944578068891

- Validation Set Error: 0.0252

39

random_seed =

- Intercept (β0): -11.298730031895952
- Coefficient for income (β1): 2.1935029663275324e-05
- Coefficient for balance (β2): 0.0054921467673926284

- Validation Set Error: 0.0236

84:

random_seed = 

- Intercept (β0): -11.267186982482086
- 
Coefficient for income (β1):2.0696240810590906e-05
- 
Coefficient for balance (β2)0.005470454917381068

-  Validation Set Error: 0.0262

68The logistic regression model shows stable performance across different random splits, with similar coefficients and validation errors. This indicates that the model generalizes well, and the dataset is balanced and representative.





124:



Question D: Now consider a logistic regression model that predicts the probability
of default using income, balance, and a dummy variable
for student. Estimate the test error for this model using the validation
set approach. Comment on whether or not including a
dummy variable for student leads to a reduction in the test error
rate.

In [8]:
# Question D: 

# Add 'student' as a predictor in the model
X = default_data[['income', 'balance', 'student']]  # Include the 'student' variable
y = default_data['default']

# Initialize a list to store validation errors
validation_errors_with_student = []

# Perform validation set approach with seeds 44, 84, 124
for random_seed in [44, 84, 124]:
    # Split the data into training and validation sets
    X_train, X_validation, y_train, y_validation = train_test_split(X, y, test_size=0.5, random_state=random_seed)
    
    # Fit logistic regression model
    logit_model = LogisticRegression()
    logit_model.fit(X_train, y_train)
    
    # Predict on the validation set
    y_prob = logit_model.predict_proba(X_validation)[:, 1]  # Probabilities for 'default = 1'
    y_pred = (y_prob > 0.5).astype(int)  # Classify based on threshold of 0.5
    
    # Compute validation error
    num_misclassified = (y_validation != y_pred).sum()
    total_observations = len(y_validation)
    validation_error = num_misclassified / total_observations
    
    # Store the validation error
    validation_errors_with_student.append(validation_error)
    print(f"Random Seed {random_seed} - Validation Error (with student): {validation_error}")




Random Seed 44 - Validation Error (with student): 0.0264
Random Seed 84 - Validation Error (with student): 0.0242
Random Seed 124 - Validation Error (with student): 0.0272


Including the student variable increased the validation error across all three splits, suggesting it does not significantly improve the model's performance. This indicates that student may not be a strong predictor for default in this model.